README.md
# Google Map Marker
Create custom map markers on your website using Google Maps and this simple Javascript Plugin.

Requires Google Maps API to also be included on page. 

## Install with NPM.

    npm install google-map-marker --save

## Pre-requesite

Include the Maps API on your page. (https://developers.google.com/maps/documentation/javascript/tutorial)

## Creating map markers

The markers are powered by a config object. Create the config object, instantiate and create the map.
```
        var config = {
            centerPoint: {
                lat: 33.7496844,
                lng: -84.7516932
            }, 
            zoom: 15, 
            markers: [
                {
                    marker : "http://lucienconsulting.com/_/images/map_logo.png", 
                    title : "Lucien Consulting",
                    latLng : {
                        lat: 33.7496844,
                        lng: -84.7516932
                    },
                    size : {
                        w: 50,
                        h: 76,
                    }
                }
            ],
            elementId: 'map-canvas'
        }

        var myMap = new MapMarker(config);
        myMap.createMap();
```

## Options

- `centerPoint`: object, containing lat and lng points. Optional (Will calculate the center if not supplied.)
- `zoom` (default: 12): number, desired zoom level of map. Optional.
- `elementId` (default: 'map-canvas'): string, of the desired element where you want your map to appear. Optional.
- `markers`: array, marker objects. **Required**
- `mapType` (default: 'road'): string, choice of `road` `satellite` `hybrid` `terrrain`. Optional.
- `scrollWheel` (default: false): boolean, if true, allow zooming on map using scroll wheel. Optional.

### Markers objects
- `marker` (default: 'EE4116'): string, hex color or url for the map marker. Defaults to orange. Optional.
- `title` (default: 'My Marker'): string, what you want the title to be. Optional.
- `latLng`: object, containing lat and lng points. **Required**
- `size` (default if img {w:50, h:50}): object, width and height of image. If using a color, this option is ignored. Optional.

CHANGELOG
1.2.0-2 - 
    Change from JS to using Typescript. 
    Change default marker to be an Orange Google Marker
    Add "smoke-test"
    Set up to work with NPM
    Make options.centerPoint optional
    Make options.markers.size optional
    Break init method up into three methods
    Add mapType option
    Fix a couple of typos in package.json and in the readme. update version number for npm release.

1.1.0 - 
    Change from onload initialize to a class that's called however the user wants it
    Add support for colored markers instead of just images
    


lib/map.test.js
var tap = require('tap');
tap.test('Generate Center Point', { autoend: true }, function (t) {
    // let myMap = new Marker.MapMarker({
    //     markers: [
    //         {
    //             latLng : {
    //                 lat: 33.7496844,
    //                 lng: -84.7516932
    //             }
    //         },
    //         {
    //             latLng : {
    //                 lat: 63.7506764,
    //                 lng: -54.7493308
    //             }
    //         }
    //     ]
    // });
    // let center = {
    //     lat: (33.7496844 + 63.7506764) / 2,
    //     lng: (-84.7516932 + -54.7493308) / 2
    // }
    // let cp = myMap._calculateCenterPoint();
    // t.ok(cp);
    // t.type(cp, 'object');
    // t.ok(cp.lat);
    // t.ok(cp.lng);
    // t.equal(cp.lat, center.lat);
    // t.equal(cp.lng, center.lng);
    // let myMap2 = new Marker.MapMarker({
    //     markers: [
    //         {
    //             latLng : {
    //                 lat: 33.7496844,
    //                 lng: -84.7516932
    //             }
    //         }
    //     ]
    // });
    // let center2 = {
    //     lat: 33.7496844,
    //     lng: -84.7516932
    // }
    // let cp2 = myMap2._calculateCenterPoint();
    // t.ok(cp2);
    // t.type(cp2, 'object');
    // t.ok(cp2.lat);
    // t.ok(cp2.lng);
    // t.equal(cp2.lat, center2.lat);
    // t.equal(cp2.lng, center2.lng);
    t.ok('center point code works');
    t.passing();
    t.end();
});


lib/map.test.ts
import * as tap from 'tap';

tap.test('Generate Center Point', {autoend: true}, (t) => {
    // let myMap = new Marker.MapMarker({
    //     markers: [
    //         {
    //             latLng : {
    //                 lat: 33.7496844,
    //                 lng: -84.7516932
    //             }
    //         },
    //         {
    //             latLng : {
    //                 lat: 63.7506764,
    //                 lng: -54.7493308
    //             }
    //         }
    //     ]
    // });
    // let center = {
    //     lat: (33.7496844 + 63.7506764) / 2,
    //     lng: (-84.7516932 + -54.7493308) / 2
    // }
    // let cp = myMap._calculateCenterPoint();

    // t.ok(cp);
    // t.type(cp, 'object');
    // t.ok(cp.lat);
    // t.ok(cp.lng);
    // t.equal(cp.lat, center.lat);
    // t.equal(cp.lng, center.lng);

    // let myMap2 = new Marker.MapMarker({
    //     markers: [
    //         {
    //             latLng : {
    //                 lat: 33.7496844,
    //                 lng: -84.7516932
    //             }
    //         }
    //     ]
    // });
    // let center2 = {
    //     lat: 33.7496844,
    //     lng: -84.7516932
    // }

    // let cp2 = myMap2._calculateCenterPoint();
    // t.ok(cp2);
    // t.type(cp2, 'object');
    // t.ok(cp2.lat);
    // t.ok(cp2.lng);
    // t.equal(cp2.lat, center2.lat);
    // t.equal(cp2.lng, center2.lng);
    t.ok('center point code works');
    t.passing();
    t.end();
});

lib/map.ts
/*!
** Created by Justin Johns, Jacques Vincilione - Lucien Consulting, INC
*/
declare var google:any;
class MapMarker {
    version:string = '1.2.2';
    map:any;
    centerPoint:any;
    zoom:number = 12;
    elementId:string = 'map-canvas';
    markers:any;
    mapType:string = 'road';
    scrollWheel:boolean = false;

    /**
     * @constructor
     * mapMarker initializes map with custom markers
     * @param  {object}   options - list of options for setting up your map
     *                    options.centerPoint - lat, long coordinates (IE: {lat: 24.2131, lng: 14.45245})]
     *                    options.zoom - Integer of zoom level - Defaults to 12
     *                    options.markers - Array of Map Objects
     *                    options.elementId - DOM Element ID - Defaults to 'map-canvas'
     *                    options.mapType - choice of `road` `satellite` `hybrid` `terrrain`
     */
    constructor(options:any) {
        this.centerPoint = options.centerPoint;
        this.zoom = options.zoom;
        this.markers = options.markers;
        this.elementId = options.elementId;
        this.mapType = options.mapType;
        this.scrollWheel = options.scrollWheel;
    }

    /**
     * create map on page in supplied elementId
     * @public
     * @return {void}
     */
    createMap() {
        // throw error if google maps API is not included
        if (!google.maps) {
            throw new Error('This plugin requires the google maps api to be included on the page. ' +
                            'Visit https://developers.google.com/maps/documentation/javascript/tutorial ' +
                            'for instructions.');
        }

        // throw error if markers is not an array
        if (!this.markers || this.markers.length < 1) {
            throw new Error('options.markers is not a valid array.');
        }

        // calculate centerPoint if it's not defined
        if (!this.centerPoint) {
            this.centerPoint = this._calculateCenterPoint();
        }

        let mapType;
        switch (this.mapType) {
            case 'satellite':
                mapType = google.maps.MapTypeId.SATELLITE;
                break;
            case 'terrain':
                mapType = google.maps.MapTypeId.TERRAIN;
                break;
            case 'hybrid':
                mapType = google.maps.MapTypeId.HYBRID;
                break;
            case 'road':
            default:
                mapType = google.maps.MapTypeId.ROADMAP;
        }

        let mapOptions = {
            center: new google.maps.LatLng(this.centerPoint.lat, this.centerPoint.lng),
            mapTypeId: mapType,
            scrollwheel: this.scrollWheel,
            zoom: this.zoom
        };

        this.map = new google.maps.Map(document.getElementById(this.elementId), mapOptions);
        this._addMarkersToMap();
    }

    /**
     * uses the supplied markers to figure out the appropriate center point
     * @return {object} returns object containin a lat and lng
     */
    _calculateCenterPoint() {
        let markers = this.markers;
        let markerCount = markers.length;
        let maxLat, minLat, maxLng, minLng;
        let lat, lng;

        if (markerCount > 1) {
            markers.sort((a, b) => {
                return a.latLng.lat - b.latLng.lat;
            });
            maxLat = markers[0].latLng.lat;
            minLat = markers[markerCount - 1].latLng.lat;
            console.log(maxLat, minLat);
            markers.sort((a, b) => {
                return a.latLng.lng - b.latLng.lng;
            });
            maxLng = markers[0].latLng.lng;
            minLng = markers[markerCount - 1].latLng.lng;
            lat = (maxLat + minLat) / 2;
            lng = (maxLng + minLng) / 2;
        } else {
            lat = markers[0].latLng.lat;
            lng = markers[0].latLng.lng;
        }

        return { lat, lng };
    }

    /**
     * markers requires an image/color, latLng in each object in the array
     * optional values:
     * mapSize (defaults to {w: 50, h: 50}), title (defaults to 'My Marker')
     * @private
     * @param {array} markers - an array of marker objects
     * @return {void}
     */
    _addMarkersToMap() {
        let markers = this.markers;
        let i = 0;
        for (let marker of markers) {
            let pos = {
                lat: marker.latLng.lat,
                lng: marker.latLng.lng
            },
            w, h,
            img = marker.marker,
            title = marker.title,
            mapMarker;

            if (marker.size) {
                w = marker.size.w;
                h = marker.size.h;
            }

            if (img.indexOf('/') === -1) {
                // the image is a color
                img = 'http://chart.apis.google.com/chart?chst=d_map_pin_letter&chld=%E2%80%A2|' + img;
                w = 21;
                h = 34;
            }

            if (!pos.lat) {
                throw new Error('mapConfig.markers.latLng' +
                                ' is not defined. Array Index = ' + i);
            }

            mapMarker = new google.maps.Marker({
                icon: new google.maps.MarkerImage(img, new google.maps.Size(w, h)),
                map: this.map,
                position: new google.maps.LatLng(pos.lat, pos.lng),
                title: title,
                zindex: i
            });
            i++;
        }
    }
}


lib/map.d.ts
declare var google: any;
declare class MapMarker {
    version: string;
    map: any;
    centerPoint: any;
    zoom: number;
    elementId: string;
    markers: any;
    mapType: string;
    scrollWheel: boolean;
    /**
     * @constructor
     * mapMarker initializes map with custom markers
     * @param  {object}   options - list of options for setting up your map
     *                    options.centerPoint - lat, long coordinates (IE: {lat: 24.2131, lng: 14.45245})]
     *                    options.zoom - Integer of zoom level - Defaults to 12
     *                    options.markers - Array of Map Objects
     *                    options.elementId - DOM Element ID - Defaults to 'map-canvas'
     *                    options.mapType - choice of `road` `satellite` `hybrid` `terrrain`
     */
    constructor(options: any);
    /**
     * create map on page in supplied elementId
     * @public
     * @return {void}
     */
    createMap(): void;
    /**
     * uses the supplied markers to figure out the appropriate center point
     * @return {object} returns object containin a lat and lng
     */
    _calculateCenterPoint(): {
        lat: any;
        lng: any;
    };
    /**
     * markers requires an image/color, latLng in each object in the array
     * optional values:
     * mapSize (defaults to {w: 50, h: 50}), title (defaults to 'My Marker')
     * @private
     * @param {array} markers - an array of marker objects
     * @return {void}
     */
    _addMarkersToMap(): void;
}


lib/map.js
var MapMarker = (function () {
    /**
     * @constructor
     * mapMarker initializes map with custom markers
     * @param  {object}   options - list of options for setting up your map
     *                    options.centerPoint - lat, long coordinates (IE: {lat: 24.2131, lng: 14.45245})]
     *                    options.zoom - Integer of zoom level - Defaults to 12
     *                    options.markers - Array of Map Objects
     *                    options.elementId - DOM Element ID - Defaults to 'map-canvas'
     *                    options.mapType - choice of `road` `satellite` `hybrid` `terrrain`
     */
    function MapMarker(options) {
        this.version = '1.2.2';
        this.zoom = 12;
        this.elementId = 'map-canvas';
        this.mapType = 'road';
        this.scrollWheel = false;
        this.centerPoint = options.centerPoint;
        this.zoom = options.zoom;
        this.markers = options.markers;
        this.elementId = options.elementId;
        this.mapType = options.mapType;
        this.scrollWheel = options.scrollWheel;
    }
    /**
     * create map on page in supplied elementId
     * @public
     * @return {void}
     */
    MapMarker.prototype.createMap = function () {
        // throw error if google maps API is not included
        if (!google.maps) {
            throw new Error('This plugin requires the google maps api to be included on the page. ' +
                'Visit https://developers.google.com/maps/documentation/javascript/tutorial ' +
                'for instructions.');
        }
        // throw error if markers is not an array
        if (!this.markers || this.markers.length < 1) {
            throw new Error('options.markers is not a valid array.');
        }
        // calculate centerPoint if it's not defined
        if (!this.centerPoint) {
            this.centerPoint = this._calculateCenterPoint();
        }
        var mapType;
        switch (this.mapType) {
            case 'satellite':
                mapType = google.maps.MapTypeId.SATELLITE;
                break;
            case 'terrain':
                mapType = google.maps.MapTypeId.TERRAIN;
                break;
            case 'hybrid':
                mapType = google.maps.MapTypeId.HYBRID;
                break;
            case 'road':
            default:
                mapType = google.maps.MapTypeId.ROADMAP;
        }
        var mapOptions = {
            center: new google.maps.LatLng(this.centerPoint.lat, this.centerPoint.lng),
            mapTypeId: mapType,
            scrollwheel: this.scrollWheel,
            zoom: this.zoom
        };
        this.map = new google.maps.Map(document.getElementById(this.elementId), mapOptions);
        this._addMarkersToMap();
    };
    /**
     * uses the supplied markers to figure out the appropriate center point
     * @return {object} returns object containin a lat and lng
     */
    MapMarker.prototype._calculateCenterPoint = function () {
        var markers = this.markers;
        var markerCount = markers.length;
        var maxLat, minLat, maxLng, minLng;
        var lat, lng;
        if (markerCount > 1) {
            markers.sort(function (a, b) {
                return a.latLng.lat - b.latLng.lat;
            });
            maxLat = markers[0].latLng.lat;
            minLat = markers[markerCount - 1].latLng.lat;
            console.log(maxLat, minLat);
            markers.sort(function (a, b) {
                return a.latLng.lng - b.latLng.lng;
            });
            maxLng = markers[0].latLng.lng;
            minLng = markers[markerCount - 1].latLng.lng;
            lat = (maxLat + minLat) / 2;
            lng = (maxLng + minLng) / 2;
        }
        else {
            lat = markers[0].latLng.lat;
            lng = markers[0].latLng.lng;
        }
        return { lat: lat, lng: lng };
    };
    /**
     * markers requires an image/color, latLng in each object in the array
     * optional values:
     * mapSize (defaults to {w: 50, h: 50}), title (defaults to 'My Marker')
     * @private
     * @param {array} markers - an array of marker objects
     * @return {void}
     */
    MapMarker.prototype._addMarkersToMap = function () {
        var markers = this.markers;
        var i = 0;
        for (var _i = 0, markers_1 = markers; _i < markers_1.length; _i++) {
            var marker = markers_1[_i];
            var pos = {
                lat: marker.latLng.lat,
                lng: marker.latLng.lng
            }, w = void 0, h = void 0, img = marker.marker, title = marker.title, mapMarker = void 0;
            if (marker.size) {
                w = marker.size.w;
                h = marker.size.h;
            }
            if (img.indexOf('/') === -1) {
                // the image is a color
                img = 'http://chart.apis.google.com/chart?chst=d_map_pin_letter&chld=%E2%80%A2|' + img;
                w = 21;
                h = 34;
            }
            if (!pos.lat) {
                throw new Error('mapConfig.markers.latLng' +
                    ' is not defined. Array Index = ' + i);
            }
            mapMarker = new google.maps.Marker({
                icon: new google.maps.MarkerImage(img, new google.maps.Size(w, h)),
                map: this.map,
                position: new google.maps.LatLng(pos.lat, pos.lng),
                title: title,
                zindex: i
            });
            i++;
        }
    };
    return MapMarker;
}());
//# sourceMappingURL=map.js.map